src/main/kotlin/com/grepp/nbe1_3_team04/chat/service/response/ChatResponse.kt

```diff
@@ -0,0 +1,26 @@
+package com.grepp.nbe1_3_team04.chat.service.response
+
+import com.grepp.nbe1_3_team04.chat.domain.Chat
+import com.grepp.nbe1_3_team04.chat.domain.ChatType
+import java.time.LocalDateTime
+
+@JvmRecord
+data class ChatResponse(
+    val chatId: Long?,
+    val createdAt: LocalDateTime?,
+    val updatedAt: LocalDateTime?,
+    val chatroomResponse: ChatroomResponse,
+    val memberInfo: ChatMemberInfo,
+    val chatType: ChatType,
+    val text: String
+) {
+    constructor(chat: Chat) : this(
+        chat.chatId,
+        chat.createdAt,
+        chat.updatedAt,
+        ChatroomResponse(chat.chatRoom),
+        ChatMemberInfo(chat.member),
+        chat.chatType,
+        chat.text
+    )
+}
```

Comment:
응답들 쭉보면서 느낀건데 ID는 왜 다 Nullable하게 해두는건가요 ? 여기서 시간에 대해서도 nullable한게 궁금해여

Reply:
엔티티 클래스에서 (chatId) id값이 nullable하다보니 아래 chat을 매개변수로 생성자를 만들어줄때 chatId가 null이 들어갈 가능성이 있어서 이렇게 만들었습니다! 저렇게 하지 않으면 chatId가 null일 때 따로 초기화 시켜줘야 하는데 임의값으로 초기화시켜주기보단 nullable하게 만들어주는게 좋을 것 같아서 이렇게 작성했어요!

혹시 더 좋은 의견이 있으시다면 말씀해주시면 감사하겠습니다!

Reply:
제 생각에는 응답으로 나가는 ID값이 null이면 안된다고 생각하는데 아닌가요..?

DTO는 Data Transfer Object 즉 데이터 전달만의 역할을 하기 때문에 null 일 이유가 없다고 생각합니다 !
실제로 코틀린 코드에서도 보면, 이게 어떤 방식인지 모르겠지만 어쨋든 `constructor` 키워드 자체가 생성자를 통해서 응답을 만들 것으로 보여집니다.
여기서 매개변수로 받는 Chat의 ChatID가 null일 가능성도 없을 뿐더러, 만약 ChatID가 null이라면 오히려 문제가 있는 상황이기 때문에 `?`를 붙이기 보다는 널을 허용하지 않는 편이 좋지 않을까 생각합니다.

대개 서비스 로직이 흘러가는 흐름이
```java
// 요청을 통한 처리
// 데이터베이스를 통한 작업
// 응답 생성
```

이렇게 흘러간다고 생각하는데 사실 ChatID가 null이려면 데이터베이스를 통한 작업쪽에서 무언가 문제가 발생되어야 하는거 아닐까요 ?

가장 간단하게 조회하는 로직을 자바로 생각해본다면

```java
Long id = request.getId()
Chat chat = chatRepository.findById(id);
ChatResponse response = ChatResponse.from(Chat)
```

대충 이런 흐름이라고 생각하면 응답을 담당하는 DTO에서는 id에는 널을 불가하게 하고 오히려 NULL이 들어오는게 더 예외상황이라고 판단하는게 좋지 않을까요 ?

저도 코틀린에 대해서는 잘 몰라서 자바관점에서 생각해본걸 코틀린으로 옮겨본 것이라서 코틀린에 존재하는 어떤 규칙이나, 자바와 다른점이 있다면 알려주시면 감사하겠습니다

Reply:
민혁님 말씀이 맞습니다! 다만 id에 대한 프로퍼티가 nullable하게 설정되어 있기 때문에 코틀린에서는 nullable하지 않은 프로퍼티에 null 값이 들어오는걸 막으려고 컴파일 에러가 납니다.

저 코드로 볼 때 생성자에 필요한 매개변수 chat은 서비스쪽에서 검증이 한번 된 엔티티라고 생각하고 nullable하게 두었는데 민혁님 말씀대로 DTO에서는 id를 꼭 전달해야 하니 null을 허용하지 않아야겠네요

그러면 생성자에서 chatId가 null일 경우 예외를 던지는 식으로 변경하는게 좋을 것 같은데 어떻게 생각하시나요?

Reply:
제가 코틀린을 잘 이해를 못해서 그런지..? 이런 상황이라는건가요 ?
```suggestion
data class ChatResponse(
    val chatId: Long,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime,
    val chatroomResponse: ChatroomResponse,
    val memberInfo: ChatMemberInfo,
    val chatType: ChatType,
    val text: String
) {
    constructor(chat: Chat) : this(
        chat.chatId,
        chat.createdAt,
        chat.updatedAt,
        ChatroomResponse(chat.chatRoom),
        ChatMemberInfo(chat.member),
        chat.chatType,
        chat.text
    )
}
```

이렇게 하면 컴파일 오류가 난다 이런뜻인가요 ?

이렇게 했을 때 컴파일 오류가 난다면 어쩔수없다고 생각하긴합니다... 예외처리를 하든 `!!`를 이용해서 단언을 하던 해야될 것 같아요

`!!` 쓰는게 단언하는 것이다 보니 안티패턴인지는 모르겠는데... 
제가 생각하기에 아무도 모르는 사람이 ChatResponse라는 객체를 열어봤을 때 `!!`를 통해서 "절대 NULL아님 !" 이라고 보여지는건 어떨까요? 예외처리하는 로직도 나쁘진않아보이는데... `!!` 을 사용하는 경우가 대부분 절대 NULL일수가 없는 상황에서 사용한다고는 하니까, 위험한 방식일수도 있긴 하지만 응답에서는 사용하는것도 괜찮지 않을까요 ?
 실제로는 예외가 발생할 일이 0에 수렴함에도 불구하고 계속 예외처리를 해줘야돼서 귀찮아 질 것 같아요..

Reply:
저도 찬성합니다.

```kotlin
constructor(chat: Chat) : this(
        chatId = requireNotNull(chat.chatId) { "chatId는 null이 될 수 없습니다." },
        createdAt = chat.createdAt,
        updatedAt = chat.updatedAt,
        chatroomResponse = ChatroomResponse(chat.chatRoom),
        memberInfo = ChatMemberInfo(chat.member),
        chatType = chat.chatType,
        text = chat.text
    )
 ```
이런 식으로 작성하면 될까요?

Reply:
민혁님이 작성해주신 코드가 맞습니다! 저렇게 작성하면 컴파일오류가 납니다.

`!!` 를 쓰는건 최대한 지양해야 한다고 합니다. 되도록 엘비스 연산자를 사용하는게 좋다고 하네요

엘비스 연산자 사용해서 예외처리를 해주거나 관태님이 작성해주신 코드처럼 작성하는게 좋아 보입니다!

관태님 코드가 훨씬 깔끔하니 저렇게 하면 좋을거같네요!! requireNotNull 메서드가 IllegalArgumentException을 던져서 저희 프로젝트에도 잘 맞을 것 같아요

추가적으로 createdAt이나 updatedAt처럼 id가 아닌 nullable한 필드는 실제로 null로 존재할 수 있는데 어떻게 하는게 좋을것같나요??

Reply:
근데 저도 이게 궁금하긴한데 실제로 nullable 할 수 없잖아요 
Auditing을 통해서 자동으로 들어가기 때문에 이것도 Auditing을 적용하기 위해서 ? 해두고 = null로 초기화 해둔 것 같은데 Kotlin에서 어떻게 해야하는지 잘 몰라서...
제 생각은 실제론 nullalbe 하지 않기 때문에 requireNotNull 하는게 좋아보입니다 ~

Reply:
넵!!! 마지막으로 description같이 null이 허용된 것도 있는데 이런건 null을 허용해주는게 맞을까요???

자바 코드에서는 따로 처리를 안한것같아서 질문드립니다! 기본값으로 초기화를 해준다거나 null값은 보내지 않는다거나 하는 처리가 필요할 것 같아요

다같이 의견 나눠보면 좋을거같습니다

Reply:
null 인채로 보내고있을겁니다 실제 동작하는거는요 응답이 json 형식이 내려가고 있기 때문에 실제로 프론트가 데이터를 받아서 사용하기 때문에 null이어도 그냥 null을 저는 보내기는 합니다.

실제로 Java에서 JsonInclude(notnull) ? 이거였나 ? 그거하면 null인 필드는 안보내는 방법도 있어요